File: takengine/mapengine/sdk/src/renderer/raster/tilematrix/GLTiledLayerCore.cpp
#include "renderer/raster/tilematrix/GLTiledLayerCore.h"

#include "math/Point2.h"
#include "core/GeoPoint2.h"
#include "core/ProjectionFactory2.h"
#include "core/ProjectionFactory3.h"
#include "math/Utils.h"

using namespace TAK::Engine::Renderer::Raster::TileMatrix;

namespace {
    TAK::Engine::Core::Projection2Ptr getProjection(int srid) {
        TAK::Engine::Core::Projection2Ptr value(nullptr, nullptr);

        // Partial port of Java's MobileImageryRasterLayer2.getProjection(srid);
        // based on input from Chris
        switch (srid) {
            case 900913:
                srid = 3857;
                break;
            case 90094326:
                srid = 4326;
                break;
        }
        ProjectionFactory3_create(value, srid);
        return value;
    }

    class ProjectionDatasetProjection2 : public atakmap::raster::DatasetProjection {
        GLTiledLayerCore *core;

      public:
        ProjectionDatasetProjection2(GLTiledLayerCore *core) : core(core) { }

        void release() {
        }

        bool imageToGround(const atakmap::math::PointD &image, atakmap::core::GeoPoint *ground) const override {
            TAK::Engine::Core::GeoPoint2 gp2;
            TAK::Engine::Math::Point2<double> p2(image.x, image.y, image.z);
            core->proj->inverse(&gp2, p2);
            *ground = atakmap::core::GeoPoint(gp2);
            return true;
        }

        bool groundToImage(const atakmap::core::GeoPoint &ground, atakmap::math::PointD *image) const override {
            TAK::Engine::Core::GeoPoint2 gp2;
            atakmap::core::GeoPoint_adapt(&gp2, ground);
            TAK::Engine::Math::Point2<double> p2;
            core->proj->forward(&p2, gp2);
            *image = atakmap::math::PointD(p2.x, p2.y, p2.z);
            return true;
        }
    };

}


GLTiledLayerCore::GLTiledLayerCore(const std::shared_ptr<TAK::Engine::Raster::TileMatrix::TileMatrix> &matrix, const char *uri)
    : 
      clientSourceUri(uri), proj(nullptr, nullptr), 
      debugDraw(false), textureCache(nullptr), bitmapLoader(nullptr), r(1.0f), g(1.0f),
      b(1.0f), a(1.0f), fullExtentMinLat(0), fullExtentMinLng(0), fullExtentMaxLat(0),
      fullExtentMaxLng(0),
      proj2geo(new ProjectionDatasetProjection2(this)),
      matrix(matrix),
      fullExtent(),
      refreshInterval(0),
      tileDrawVersion(0), lastRefresh(0)
{
    proj = getProjection(this->matrix->getSRID());
    this->matrix->getBounds(&fullExtent);
        
    Math::Point2<double> scratchD(0.0, 0.0, 0.0);
    Core::GeoPoint2 scratchG;

    scratchD.x = fullExtent.minX;
    scratchD.y = fullExtent.minY;
    proj->inverse(&scratchG, scratchD);
    double lat0 = scratchG.latitude;
    double lng0 = scratchG.longitude;
    scratchD.x = fullExtent.minX;
    scratchD.y = fullExtent.maxY;
    proj->inverse(&scratchG, scratchD);
    double lat1 = scratchG.latitude;
    double lng1 = scratchG.longitude;
    scratchD.x = fullExtent.maxX;
    scratchD.y = fullExtent.maxY;
    proj->inverse(&scratchG, scratchD);
    double lat2 = scratchG.latitude;
    double lng2 = scratchG.longitude;
    scratchD.x = fullExtent.maxX;
    scratchD.y = fullExtent.minY;
    proj->inverse(&scratchG, scratchD);
    double lat3 = scratchG.latitude;
    double lng3 = scratchG.longitude;
        
    fullExtentMinLat = atakmap::math::min(lat0, lat1, lat2, lat3);
    fullExtentMinLng = atakmap::math::min(lng0, lng1, lng2, lng3);
    fullExtentMaxLat = atakmap::math::max(lat0, lat1, lat2, lat3);
    fullExtentMaxLng = atakmap::math::max(lng0, lng1, lng2, lng3);
        
    lastRefresh = Port::Platform_systime_millis();
}

GLTiledLayerCore::~GLTiledLayerCore()
{
    delete proj2geo;
}
    
void GLTiledLayerCore::requestRefresh() {
    tileDrawVersion++;
}

void GLTiledLayerCore::drawPump() {
    if(refreshInterval <= 0L)
        return;
    // if the refresh interval has elapsed since the last refresh, bump the
    // version
    int64_t currentTime = Port::Platform_systime_millis(); 
    if((currentTime - lastRefresh) > refreshInterval) {
        tileDrawVersion++;
        lastRefresh = Port::Platform_systime_millis();
    }
}

File: takkernel/engine/src/main/java/com/atakmap/map/layer/raster/tilematrix/opengl/GLTiledLayerCore.java
package com.atakmap.map.layer.raster.tilematrix.opengl;

import com.atakmap.android.maps.graphics.GLBitmapLoader;
import com.atakmap.coremap.maps.coords.GeoPoint;

import com.atakmap.map.layer.feature.geometry.Envelope;
import com.atakmap.map.layer.raster.DatasetProjection2;
import com.atakmap.map.layer.raster.mobileimagery.MobileImageryRasterLayer2;
import com.atakmap.map.layer.raster.tilematrix.TileMatrix;
import com.atakmap.map.projection.Projection;
import com.atakmap.math.MathUtils;
import com.atakmap.math.PointD;
import com.atakmap.opengl.GLTextureCache;

/**
 * The core data structure for a Tiled Map Layer, containing properties about
 * the layer that will be utilized through the renderer infrastructure.
 * 
 * @author Developer
 *
 */
public class GLTiledLayerCore {
    /**
     * The base URL of the layer.
     */
    public final String clientSourceUri;
    /**
     * The {@link Projection} of the layer
     */
    public final Projection proj;
    
    /**
     * If <code>true</code>, renderers should perform additional drawing
     */
    public boolean debugDraw;
    
    /**
     * A cache that may be used by renderers to store texture data that has gone
     * out of view.  May be <code>null</code> if no cache is available.
     */
    public GLTextureCache textureCache;
    /**
     * The asynchronous bitmap loader to be used to service tile download
     * requests.
     */
    public GLBitmapLoader bitmapLoader;
    
    /**
     * The red color component that should be applied when rendering the layer.
     */
    public float r = 1f;
    /**
     * The green color component that should be applied when rendering the layer.
     */
    public float g = 1f;
    /**
     * The blue color component that should be applied when rendering the layer.
     */
    public float b = 1f;
    /**
     * The alpha color component that should be applied when rendering the layer.
     */
    public float a = 1f;
    
    /**
     * The minimum latitude of the full extent of the layer.
     */
    public final double fullExtentMinLat;
    /**
     * The minimum longitude of the full extent of the layer.
     */
    public final double fullExtentMinLng;
    /**
     * The maximum latitude of the full extent of the layer.
     */
    public final double fullExtentMaxLat;
    /**
     * The maximum longitude of the full extent of the layer.
     */
    public final double fullExtentMaxLng;
    
    /**
     * Converts between the projected coordinate space of the layer and
     * latitude/longitude.
     */
    public final DatasetProjection2 proj2geo;
    
    /**
     * Tile access for the layer.
     */
    public final TileMatrix matrix;
    
    /**
     * The bounds of the tile containing region, in the projected coordinate
     * space of the layer.
     */
    public final Envelope fullExtent;
    
    public long refreshInterval;
    private long lastRefresh;

    public int tileDrawVersion;

    /**
     * Creates a new <code>GLTiledLayerCore</code>
     * 
     * @param service       The data structure describing the service
     * @param serviceUrl    The base URL of the map server
     */
    public GLTiledLayerCore(TileMatrix matrix, String uri) {
        this.clientSourceUri = uri;
        this.proj = MobileImageryRasterLayer2.getProjection(matrix.getSRID());
        
        this.matrix = matrix;
        this.fullExtent = this.matrix.getBounds();
        
        PointD scratchD = new PointD(0d, 0d, 0d);
        GeoPoint scratchG = GeoPoint.createMutable();

        scratchD.x = this.fullExtent.minX;
        scratchD.y = this.fullExtent.minY;
        this.proj.inverse(scratchD, scratchG);
        double lat0 = scratchG.getLatitude();
        double lng0 = scratchG.getLongitude();
        scratchD.x = this.fullExtent.minX;
        scratchD.y = this.fullExtent.maxY;
        this.proj.inverse(scratchD, scratchG);
        double lat1 = scratchG.getLatitude();
        double lng1 = scratchG.getLongitude();
        scratchD.x = this.fullExtent.maxX;
        scratchD.y = this.fullExtent.maxY;
        this.proj.inverse(scratchD, scratchG);
        double lat2 = scratchG.getLatitude();
        double lng2 = scratchG.getLongitude();
        scratchD.x = this.fullExtent.maxX;
        scratchD.y = this.fullExtent.minY;
        this.proj.inverse(scratchD, scratchG);
        double lat3 = scratchG.getLatitude();
        double lng3 = scratchG.getLongitude();
        
        this.fullExtentMinLat = MathUtils.min(lat0, lat1, lat2, lat3);
        this.fullExtentMinLng = MathUtils.min(lng0, lng1, lng2, lng3);
        this.fullExtentMaxLat = MathUtils.max(lat0, lat1, lat2, lat3);
        this.fullExtentMaxLng = MathUtils.max(lng0, lng1, lng2, lng3);
        
        this.proj2geo = new ProjectionDatasetProjection2(this.proj);
        
        this.lastRefresh = System.currentTimeMillis();
    }
    
    public void requestRefresh() {
        tileDrawVersion++;
    }

    public void drawPump() {
        if(refreshInterval <= 0L)
            return;
        // if the refresh interval has elapsed since the last refresh, bump the
        // version
        final long currentTime = System.currentTimeMillis(); 
        if((currentTime-this.lastRefresh) > this.refreshInterval) {
            this.tileDrawVersion++;
            this.lastRefresh = System.currentTimeMillis();
        }
    }

    private static class ProjectionDatasetProjection2 implements DatasetProjection2 {

        private final Projection impl;

        public ProjectionDatasetProjection2(Projection impl) {
            this.impl = impl;
        }

        @Override
        public void release() {}

        @Override
        public boolean imageToGround(PointD image, GeoPoint ground) {
            this.impl.inverse(image, ground);
            return true;
        }

        @Override
        public boolean groundToImage(GeoPoint ground, PointD image) {
            this.impl.forward(ground, image);
            return true;
        }
        
    }
}
